[RFC - DO NOT MERGE] Abi 1 1

shefty opened the pull request:
There are two quick patches to discuss.  Both are intended to target an ABI update release at the end of the year.

The first defines how to report the fabric level source address data to an app.  This change can actually go in anytime after providers have been updated to comply with the new behavior.  We need to decide if the defined behavior is what we actually want.  The concept of using the apps's requested version to select the behavior that is introduced by this change will apply to future patches.

For example, new fields will be added to fi_info.  The fi_getinfo version will indicate if the app knows how to make use of those new fields.  This change handles apps coded to an older version of libfabric, but compiled against a newer library.

The second patch does not work, but is intended to show where the ABI will change and how backwards compatibility could be supported. This change handles apps linked against an older version of libfabrc, but running with a newer version.  The problem with the change is that attempting to run code compiled against the older version of libfabric results in this error:

fi_dom_test: relocation error: fi_dom_test: symbol fi_fabric, version FABRIC_1.0 not defined in file libfabric.so.1 with link time reference

I can fix this by wrapping fi_fabric and the other 1.0 ABI calls similar to what was done for fi_getinfo.  But I'm hoping someone with more knowledge of symver can suggest an alternative that doesn't require wrapping all existing calls.

Finally, we can debate whether these changes should result in the ABI bumping to 1.1 or 2.0.

shefty commented:
@jsquyres @goodell @bturrubiates @jithinjosepkl @sayantansur @j-xiong @a-ilango @sungeunchoi @hppritcha  -- Did I miss anyone?

sayantansur commented:
> When a domain has been configured with FI_SOURCE, completions
should return the fi_addr_t associated with the source address
for a received message.  However, it is possible that no fi_addr_t
value is available.  For example, the source address has not been
inserted into the corresponding AV.

Are there other situations possible when the fi_addr_t is not available? i.e. not only the situation where the source address has not been inserted into the corresponding AV?

My concern is that there may be a memory cost for implementing a lookup feature to go from raw network address to fi_addr_t. When there are millions of addresses, this could be non-trivial in terms of memory, if not lookup time. This will be the case where fi_addr_t is an encoded 64-bit value (not a pointer to a blob).

Does a provider have an 'out' for always just delivering the raw network address? Some apps might be OK with looking at info::addr_format and interpreting things.

shefty commented:
> Are there other situations possible when the fi_addr_t is not available?
> i.e. not only the situation where the source address has not been inserted
> into the corresponding AV?

I can't think of other situations.

> My concern is that there may be a memory cost for implementing a lookup
> feature to go from raw network address to fi_addr_t.

That cost exists today when FI_SOURCE is specified.

> millions of addresses, this could be non-trivial in terms of memory, if
> not lookup time. This will be the case where fi_addr_t is an encoded 64-
> bit value (not a pointer to a blob).

> Does a provider have an 'out' for always just delivering the raw network
> address? Some apps might be OK with looking at info::addr_format and
> interpreting things.

This is what we need to discuss.  I thought of a couple of options.

- If FI_SOURCE is specified, the provider must either return a valid fi_addr_t (one the app knows about) or the raw source data.  This is what I defined here.  The app opts-in to this behavior by selecting version 1.x.

- Define a different capability bit or CQ attribute flag, so that the app must request this specific behavior.

When using FI_AV_MAP, there is always the possibility that the app will use an fi_addr_t value that will work, but is for an address that was never inserted into the local AV.  We don't want to enforce that the fi_addr_t on a transfer truly came from an fi_av_insert call because of the overhead, but it's not something that an app should rely on.  It sounds like you're describing a situation where the provider can return a valid fi_addr_t that was never inserted.  I agree that the provider should not be forced to validate that the address was inserted when no mapping is required.  The fi_addr_t in that case should be usable by the app.

Related to this is the provider behavior.  Rather than reporting an error, we could force the provider to insert the address into the AV and only report a completion failure if that does not succeed.  But I don't like the idea that unsolicited messages from an unknown source should automatically fill up the AV, or that the completion processing should be deferred while the insertion completes.

So... for FI_AV_INDEX, I can't think of a compelling alternative to this general approach.  (We can debate how an app opts-into the behavior.)  FI_AV_MAP has a different problem in that the address may not be validated against the AV on either transmits or receives, for performance reasons.  We could document this behavior.

pmmccorm commented:
Regarding the symbol versioning and ABI:

I think it can be simpler.

For the case where we are adding fields to fi_info: this is a ABI breaking change since fi_info is a part of our exported 1.0 ABI (returned by fi_getinfo). This means we bump ABI (and probably version) to 2.0.

A major (x.0) ABI/version move means users of libfabric need to recompile their apps and may need to change their code-- if they don't then we only need to bump the minor ABI/version number.

Inside libfabric this should means:

1) Add fields to fi_info struct.
2) Make a new fi_getinfo function: fi_getinfo2, mark it as ABI 2 and default linking symbol.
3) Update all providers to implement this new getinfo call.
4) Update ABI + version and release.

Following this method: any app that is not recompiled against the new libfabric-2.0 continues to work as before, any app that is newly compiled uses the new API+ABI.

shefty commented:
Adding new fields to an existing structure with a new entry point does not work.  fi_getinfo 1.0 expects fi_info 1.0.  This is why the existing structure is copied (fi_info_1_0) and passed into a new call fi_getinfo_1_0 call.  Existing applications will only have the fi_info_1_0 structure format.  The fi_getinfo_1_0 symbol is the new entry point for fi_getinfo@FABRIC_1.0.

Applications that are simply re-compiled (think source code distribution to the end-user) will end up picking up the new fi_info definition, but call the existing function name (fi_getinfo).  That doesn't work.  The new structure definition must be redirected to an fi_getinfo entry point that is aware of the changes.

The alternative is to duplicate all entry points and structures, and append a 2 to the end of the names.  But this requires more invasive changes to the app.

sayantansur commented:
@shefty : I'm talking of the situation where the fi_addr_t is just a packed uint64_t. The provider is responsible for packing this, but theoretically, nothing prevents the app from just crafting an appropriate fi_addr_t, although that isn't something that is recommended ... just that the provider has no way of telling the two apart. In such an implementation and FI_AV_MAP, there is no table that the provider maintains, and therefore cannot check on completion.

Under such situation, implementing FI_SOURCE seems to be a pain :-) On a 64-bit system even one pointer per address doubles the memory requirement ...

pmmccorm commented:
@shefty Yes that is right. I thought more about this and I think what you did should be more or less correct-- and it is going to get complicated with various different scenarios:

- add member(s) to struct
- remove or change order of struct member(s)
- add parameter to ABI function
- remove function from ABI (not hard but we should define a deprecation/removal policy now)

Also: symbol versioning compiler+linker support may be missing in other OSes: OS-X, FreeBSD and our workaround (gcc's builtin alias) may stop working with multiple functions trying to alias the same name.

And: I am guessing external built providers will break somehow with this, just not sure how.

I think all the versioning details once figured out should be collected in a document (VERSIONING.md).

shefty commented:
Luckily other OS's aren't a problem at the moment.  We don't yet support FreeBSD or Solaris with any release, and OS X is for development purposes only.  Or at least that's our out.

But... I'm still left with the problem that what I have doesn't exactly, well, work.

shefty commented:
I updated the versioning patch.  It now works.  I had to re-work the symver macros, and I added a big comment to the header file describing the macros and how to use with, along with an example.  (Plus there's the actual libfabric code that someone could look at.)  I don't anticipate breaking the ABI often, or that the upcoming break will be too significant.  Most changes are geared toward improving the coordination between the app and providers to allow for optimal performance.
